k9mail/src/main/java/com/fsck/k9/controller/MessagingController.java

```diff
@@ -137,22 +137,30 @@
     public static class UidReverseComparator implements Comparator<Message> {
         @Override
         public int compare(Message o1, Message o2) {
-            if (o1 == null || o2 == null || o1.getUid() == null || o2.getUid() == null) {
-                return 0;
-            }
-            int id1, id2;
+            Integer id1, id2;
             try {
                 id1 = Integer.parseInt(o1.getUid());
+            } catch (NullPointerException | NumberFormatException e) {
+                id1 = null;
+            }
+            try {
                 id2 = Integer.parseInt(o2.getUid());
-            } catch (NumberFormatException e) {
+            } catch (NullPointerException | NumberFormatException e) {
+                id2 = null;
+            }
+            // noinspection NumberEquality, this checks if both are null
+            boolean bothAreNull = id1 == id2;
```

Comment:
This comparison could also be true if `id1` and `id2` are small integers where the same `Integer` instance is used from the integer cache. The method will still return the correct value, but the path taken is unexpected.

I propose to change the code to the following which is more explicit for all cases:

```java
if (id1 == null && id2 == null) {
    return 0;
} else if (id1 == null) {
    return 1;
} else if (id2 == null) {
    return -1;
}
```